lib/core/src/Cardano/Wallet/DB/Sqlite/Migration.hs

```diff
@@ -116,6 +131,67 @@ migrateManually tr proxy defaultFieldValues =
     , cleanupSeqStateTable
     ]
   where
+    currentFileFormat :: DatabaseFileFormatVersion
+    currentFileFormat = DatabaseFileFormatVersion 1
+
+    initializeDatabaseMetadataTable :: Sqlite.Connection -> IO ()
+    initializeDatabaseMetadataTable conn =
+        createMetadataTableIfMissing conn >>= \case
+            TableCreated -> do
+                trace DatabaseMetadataCreated
+                putMetadata conn DatabaseMetadata
+                    { fileFormatVersion = currentFileFormat }
+            TableExisted -> do
+                DatabaseMetadata {..} <- getMetadata conn
+                when (fileFormatVersion > currentFileFormat) do
+                    trace $ DatabaseVersionMismatched
+                        currentFileFormat -- expected
+                        fileFormatVersion -- actual
+                    let showVer = show . naturalDatabaseFileFormat
+                    throwString $ unlines
+                        [ "Unexpected database file format version:"
+                        , "- Expected version <= " <> showVer currentFileFormat
+                        , "- Actual version == " <> showVer fileFormatVersion
+                        ]
+                trace $ DatabaseVersionMatched fileFormatVersion
+                putMetadata conn DatabaseMetadata {..}
+        where trace = traceWith tr . MsgMetadata
+
+    createMetadataTableIfMissing :: Sqlite.Connection -> IO TableCreationResult
+    createMetadataTableIfMissing conn = do
+        res <- runSql conn
+            "SELECT name FROM sqlite_master \
+            \WHERE type='table' AND name='database_metadata'"
+        case res of
+           [] -> TableCreated <$ runSql conn
+            "CREATE TABLE database_metadata\
```

Comment:
> To me the `database_metadata seems` to be a better name than `file_format` because:

My view is actually that the `database` is the implementation detail and that `file` is the role — the main purpose of this thing is to persist the wallet state onto disk, and the file format just happens to be an .sqlite file for better interoperability. (Consequently, I consider the :memory: database to be useful for testing only.) That said, I don't mind the name of the table, though I would prefer `database_schema_version` over `database_metadata` indeed because we also have "transaction metadata".

>  Do you think it doesn't make sense to ship this version number without reworking migrations mechanism?

I would say yes — this version number implementation solves the issue that old wallet software needs to be able to reject new file formats (and not write to the file at all).